flocker/control/_persistence.py

```diff
@@ -21,16 +23,151 @@
+    :param bytes config: The JSON-encoded source configuration.
+
+    :return bytes: The updated JSON configuration after migration.
+    """
+    upgraded_config = config
+    current_version = source_version
+    for upgrade_version in range(source_version + 1, target_version + 1):
+        migration_method = (
+            u"configuration_v%d_v%d"
+            % (current_version, upgrade_version)
+        )
+        try:
+            migration = getattr(ConfigurationMigration, migration_method)
+        except AttributeError:
+            message = (
+                u"Unable to find a migration path for a version " +
+                unicode(source_version) + u" to version " +
+                unicode(target_version) + u" configuration. " +
+                u"No migration method exists for v" +
+                unicode(current_version) + u" to v" +
+                unicode(upgrade_version) + u"."
+            )
+            raise ConfigurationMigrationError(message)
+        upgraded_config = migration(config)
+        current_version = current_version + 1
+    return upgraded_config
+
+
+class IConfiguration(Interface):
+    """
+    An ``IConfiguration`` implementation provides a serializer and
+    deserializer for a ``Configuration`` model.
+    """
+    def serialize(config):
+        """
+        Serialize the supplied configuration model to JSON.
+
+        :param Configuration config: The configuration to serialize.
+        :return bytes: The JSON representation.
+        """
+
+    def deserialize(config):
+        """
+        Deserialize the supplied JSON to a ``Configuration`` model.
+
+        :param bytes config: The JSON configuration to deserialize.
+        :return Configuration: The configuration model.
+        """
+
 
-class _ConfigurationEncoder(JSONEncoder):
+@implementer(IConfiguration)
+class Configuration_V1(object):
+    """
+    A version 1 configuration.
+    """
+    @classmethod
+    def serialize(cls, config):
+        # Serialized v1 configs are represented as Deployment objects,
+        # not Configuration objects. This is to retain backwards
+        # compatibility.
+        return wire_encode(config.deployment)
+
+    @classmethod
+    def deserialize(cls, config):
+        return Configuration(
+            version=1,
+            deployment=wire_decode(config)
+        )
+
+
+@implementer(IConfiguration)
+class Configuration_V2(object):
+    """
+    A version 2 configuration.
+    """
+    @classmethod
+    def serialize(cls, config):
+        return wire_encode(config)
+
+    @classmethod
+    def deserialize(cls, config):
+        return wire_decode(config)
+
+
+class ConfigurationMigration(object):
+    """
+    Migrate a JSON configuration from one version to another.
+    """
+    @classmethod
+    def configuration_v1_v2(cls, config):
+        """
+        Migrate a v1 JSON configuration to v2.
+
+        :param bytes config: The v1 JSON data.
```

Comment:
If it's actually specifically upgrading JSON, why is it taking bytes? Either it shouldn't be talking about JSON since it might eventually support other on-disk formats, or maybe it should be taking output of `json.loads`?